flink-core/src/main/java/org/apache/flink/configuration/Configuration.java

```diff
@@ -239,9 +239,21 @@ public void addAllToProperties(Properties props) {
     }
 
     public void addAll(Configuration other) {
-        synchronized (this.confData) {
```

Comment:
I am not totally convinced there are no issues with this locking strategy.

If we have 2 configuration object A and B, they would always lock the other then ourselves. From object A's and B's perspective the lock order is not always the same.

Reply:
This change makes we always lock one first and then another. As you said, one thread locks the other object, then all the threads will lock the other object first.

Reply:
Yes I see that is how you have coded it, but objects should always be locked in the same order. So Object A then Object B should be locked in that order, irrespective of which thread owns them.

If each thread does its own object first, then thread 1 could lock Object A , then thread 2 lock Object B, then thread 1 would be locked out of Object B, and vice versa. 

I was thinking if the objects had unique hashes then always do the lower hash first to ensure the order was the same.

Reply:
Yes. Let me take an example. There are two objects A and B. The hash value of Object A is smaller than Object B's.
Thread 1 holds the Object A and thread 2 holds the Object B. Thread 1 will lock Object A first and then Object B. Thread 2 wants lock Object B first but it can't since the hash value of Object A is lower than Object B, so this thread still needs lock Object A first and then Object B.